collectors/feed/utils.py

```diff
@@ -0,0 +1,204 @@
+        if m:
+            m = m.groups()
+            cve_id = "%s{%s}%s" % (m[0], "..".join(m[1].split("-", 1)), m[2])
+            found_cves[idx] = cve_id
+
+    # expand {1,2,3} | {1..3}
+    all_cves = []
+    for cve in found_cves:
+        expanded_cves = enumerate(braceexpand.braceexpand(cve))
+        for index, expanded_cve in expanded_cves:
+            if index > max_cve_expand:
+                continue
+            if not re.match(r"^CVE-[12]\d{3}-\d{4,}$", expanded_cve):
+                continue
+            all_cves.append(expanded_cve)
+
+    return sorted(list(set(all_cves)))
+
+
+def find_cvss(text):
+    """Find CVSS version 3 vector in text.
+
+    Returns cleaned vector with score in Red Hat notation, e.g. score/vector.
+    If more than one vector found, returns vector with the highest score.
+    """
+    if not text:
+        return ""
+
+    found_cvss_objects = parse_cvss_from_text(text)
+
+    if not found_cvss_objects:
+        return ""
+
+    # We are interested in CVSS version 3 vectors
+    cvss3_objects = [cvss for cvss in found_cvss_objects if isinstance(cvss, CVSS3)]
+
+    if not cvss3_objects:
+        return ""
+
+    # Find CVSS with highest base score
+    result = max(cvss3_objects, key=lambda cvss: cvss.base_score)
+
+    return result.rh_vector()
+
+
+def find_cwes(text, impact=None):
+    """Find all CWEs in a block of text.
+
+    If none found, guess CWEs using common phrases in text.
+    Return a string in a format described in:
+    https://docs.prodsec.redhat.com/workflow/howtos/cwe.html#_cwe_tutorial
+    """
+    if not text:
+        return
+
+    cwes = CWE_RE_STR.findall(text)
+    cwes = set(map(str.upper, cwes))
+
+    if not cwes:
+        # Guess CWEs using common phrases in text
+        text = text.lower()
+        for phrase, cwe in PHRASE_TO_CWE_MAP.items():
+            if phrase in text:
+                cwes.add(cwe)
+
+        if "integer overflow" in text:
+            if impact == "important":
+                cwes.add("CWE-190->CWE-120")
+            else:
+                cwes.add("CWE-190")
+
+    # Return in specified format
+    if not cwes:
+        return
+    elif len(cwes) == 1:
+        return cwes.pop()
+    else:
+        cwes = sorted(cwes)
+        return "(" + "|".join(cwes) + ")"
+
+
+# TODO: Reimplement? Include somehow in requirements? Prune docstring?
+def html_to_text(html):
+    """Use elinks to convert HTML to plain text.
+
+    Other HTML->text solutions were either too heavy weight, returned ugly text, or failed to
+    parse certain parts of the HTML. (BeautifulSoup is great at scraping but not so good at
+    rendering plain text; html2text had terrible output if the parsed HTML included a table.)
+
+    Update 2018-08-27: Research was done in various options of rendering website contents as plain
+    text in bug https://bugzilla.redhat.com/show_bug.cgi?id=1558705.
+    """
+    with NamedTemporaryFile(prefix="html_to_text.") as tmpfile:
```

Comment:
can `elinks` not read from stdin? because if so this could be simplified

Reply:
Yes, it apparently can!

```
# echo "<html><body><b>hello</b> world</body></html><asdfsadf>" | elinks -no-home -dump -dump-width 80 -no-numbering -no-references
   hello world
```

Reply:
Done.